source/rules/literals.rs

```diff
@@ -298,44 +299,52 @@ fn string_nowdoc(input: &[u8]) -> Result<&[u8], Literal> {
         return Result::Error(Error::Code(ErrorKind::Custom(StringError::InvalidOpeningCharacter as u32)));
     }
 
-    let padding      = 4;
-    let mut offset   = padding;
-    let mut iterator = input[offset..].iter().enumerate();
-
-    while let Some((index, item)) = iterator.next() {
-        if *item == '\'' as u8 {
-            offset += index;
+    let name;
+    let next_input;
 
-            break;
-        }
+    if let Result::Done(i, n) = tokens::name(&input[4..]) {
+        name       = n;
+        next_input = i;
+    } else {
+        return Result::Error(Error::Code(ErrorKind::Custom(StringError::InvalidOpeningCharacter as u32)))
     }
 
-    if input[offset] != '\'' as u8 || input[offset + 1] != '\n' as u8 {
+    let next_input_length = next_input.len();
+
+    if next_input_length < 3 + name.len() ||
+       next_input[0] != '\'' as u8 ||
+       next_input[1] != '\n' as u8 {
         return Result::Error(Error::Code(ErrorKind::Custom(StringError::InvalidOpeningCharacter as u32)));
     }
 
-    let name       = &input[padding..offset];
-    let mut output = Vec::new();
-
-    iterator.next();
+    let offset       = 2;
+    let mut iterator = next_input[offset..].iter().enumerate();
 
     while let Some((index, item)) = iterator.next() {
         if *item == '\n' as u8 {
-            if !input[padding + index + 1..].starts_with(name) {
+            if !next_input[offset + index + 1..].starts_with(name) {
                 continue;
             }
 
-            offset                   = padding + index;
-            let mut lookahead_offset = offset + name.len() + 1;
+            let mut lookahead_offset = offset + index + name.len() + 1;
 
-            if input[lookahead_offset] == ';' as u8 {
+            if lookahead_offset > next_input_length {
+                return Result::Error(Error::Code(ErrorKind::Custom(StringError::InvalidClosingCharacter as u32)));
+            }
+
+            if next_input[lookahead_offset] == ';' as u8 {
                 lookahead_offset += 1;
             }
 
-            if input[lookahead_offset] == '\n' as u8 {
-                output.extend(&input[padding + name.len() + 2..offset]);
+            if lookahead_offset > next_input_length {
+                return Result::Error(Error::Code(ErrorKind::Custom(StringError::InvalidClosingCharacter as u32)));
+            }
 
-                return Result::Done(&input[lookahead_offset + 1..], Literal::String(output));
+            if next_input[lookahead_offset] == '\n' as u8 {
+                return Result::Done(
+                    &next_input[lookahead_offset + 1..],
+                    Literal::String(next_input[offset..offset + index].to_vec())
```

Comment:
`to_vec` copies items of the vector. This is not bad but we can still grab the ownership without the need to copy with [`into_vec`](https://doc.rust-lang.org/std/primitive.slice.html#method.into_vec) but it requires a `Box<[T]>`. I am trying something with it.